### PR TITLE
Prevent a gRPC connection from being fatally terminated due to ENOBUFS error

### PR DESCRIPTION
**Pull request by Vignesh2208**
Prevent a gRPC connection from being fatally terminated due to ENOBUFS error. (b/232829319)

@ctiller

### CODE REVIEW

**Review by arjunroy — CHANGES_REQUESTED** [review]
For the most part, my requested changes are just additions to comments. The actual logic looks good to me. 

Will approve after.

File: src/core/lib/iomgr/tcp_posix.cc [hunk 1]
```diff
@@ -945,14 +995,15 @@ static void tcp_read(grpc_endpoint* ep, grpc_slice_buffer* incoming_buffer,
 
 /* A wrapper around sendmsg. It sends \a msg over \a fd and returns the number
  * of bytes sent. */
-ssize_t tcp_send(int fd, const struct msghdr* msg, int additional_flags = 0) {
+ssize_t tcp_send(int fd, const struct msghdr* msg, int* saved_errno,
+                 int additional_flags = 0) {
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/iomgr/tcp_posix.cc:1055]: Done.

File: src/core/lib/iomgr/tcp_posix.cc [hunk 2]
```diff
@@ -312,7 +316,51 @@ class TcpZerocopySendCtx {
...
+      }
+    } else if (zcopy_enobuf_state_ != OMemState::OPEN) {
+      zcopy_enobuf_state_ = OMemState::OPEN;
+    }
+    return false;
+  }
+
  private:
+  enum class OMemState : int8_t {
+    OPEN,   // Everything is clear and omem is not full.
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/iomgr/tcp_posix.cc:414]: Added a state transition diagram

File: src/core/lib/iomgr/tcp_posix.cc [hunk 3]
```diff
@@ -312,7 +316,51 @@ class TcpZerocopySendCtx {
   // zerocopy is not useful for small transfers.
   size_t threshold_bytes() const { return threshold_bytes_; }
 
+  // Expected to be called by handler reading messages from the err queue.
+  // It is used to indicate that some OMem meory is now available. It returns
+  // true to tell the caller to mark the file descriptor as immediately
+  // writable.
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/iomgr/tcp_posix.cc:322]: Updated the comment

File: src/core/lib/iomgr/tcp_posix.cc [hunk 4]
```diff
@@ -312,7 +316,51 @@ class TcpZerocopySendCtx {
...
+  bool UpdateZeroCopyOMemStateAfterFree() {
+    MutexLock guard(&lock_);
+    if (is_in_write_) {
+      zcopy_enobuf_state_ = OMemState::CHECK;
+      return false;
+    } else if (zcopy_enobuf_state_ != OMemState::OPEN) {
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/iomgr/tcp_posix.cc]: Updated the code to make this more clear.

File: src/core/lib/iomgr/tcp_posix.cc [hunk 5]
```diff
@@ -312,7 +316,51 @@ class TcpZerocopySendCtx {
...
+      return true;
+    }
+    return false;
+  }
+
+  // Expected to be called by the thread calling sendmsg after the syscall
+  // invocation. is complete. If an enobuf is seen, it checks if the error
+  // handler has already run and free'ed up some OMem. It returns true
+  // indicating that the write can be attempted again immediately. Otherwise it
+  // returns false.
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/iomgr/tcp_posix.cc]: Updated the comment.

**Review by arjunroy — APPROVED** [review]

**Pull request merged**